gix-worktree/src/read.rs

```diff
@@ -0,0 +1,79 @@
+use crate::fs;
+use gix_object::Blob;
+use gix_path as path;
+use std::borrow::Cow;
+use std::fs::{read_link, File};
+use std::io;
+use std::io::Read;
+use std::path::Path;
+
+/// Error returned by [`Blob::read`] and related functions
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    ///
+    #[error("Could not convert symlink path to UTF8")]
+    IllformedUtf8,
+    ///
+    #[error("IO error while reading blob")]
+    Io(#[from] io::Error),
+}
+
+// TODO: what to do about precompose unicode and ignore case_here?
```

Comment:
precompose unicode relates to file paths and that should be done as early as possible. Some inputs even have to be sanitized on application level when they are passed in as argument. For details, I see no other way but to consult the `git` codebase.

Reply:
I am actually not sure :D I have been looking at the git codebase by grabbing around and found that they only seem to precompose cli arguments and opendir/readdir out (so untracked file detection work) but unless I misunderstood something this doesn't affect readlink.

Note that you can call `fdopen`/`File::open` both with precomposed/decomposed paths (they are equivalent form) so this doesn't matter for this function... except for symlinks. Here we need to be careful that the same paths get checked into the repo.

However I spend entirely too much time looking at the git codebase and it seems `readlink` seems to receive no special treatment. I dont' own a Mac so I can't confirm myself but either `readlink` just always returns precomposed unicode on OSX (which would be really odd considering realpath always returns decomposed unicode) or there is a bug in a git where symlinks are not stored in their precomposed form on mac (making sharing non-ascii symlinks between mac and other operating systems almost impossible). It's also possible I am missing something here but  the patchset that added uniocde precomposition doesn't even mention symlinks so it's either not needed or was overlook https://git.vger.kernel.narkive.com/OJcWG1uy/patch-v8-on-mac-os-and-precomposed-unicode